security: stack protection: Add more granularity for stack protection

ceolin opened the pull request:
Add more options for stack canaries.

Fixes: #80559

ceolin committed:
security: Additional option for stack canaries

Previously, when stack canaries were enabled, Zephyr applied this
protection to all functions. This commit introduces a new option that
allows stack canary protection to be applied selectively to specific
functions based on certain criteria.

Signed-off-by: Flavio Ceolin <flavio.ceolin@gmail.com>

ceolin committed:
security: Add option for explicit stack canaries

Add option to enable stack canaries only when explicitely
declared. It adds a new function attribute, __stack_protect, that
can be used to enable stack protection in a function.

Signed-off-by: Flavio Ceolin <flavio.ceolin@gmail.com>

rruuaanng commented on tests/kernel/mem_protect/stackprot/src/main.c:
```diff
@@ -70,6 +70,22 @@ void __attribute__((noinline)) check_input(const char *name, const char *input)
+ * so, unless it is built with stack protection for all functions it should fail
+ * with an error different from K_ERR_STACK_CHK_FAIL.
+ */
+void __attribute__((noinline)) check_input_fail(const char *name, const char *input)
```
If I remember correctly, there seems to be an `__noinline`.

ceolin replied:
Just followed what was already done in this test. I can add a commit changing the other function too.

ceolin replied:
We don't have it well defined. Just added here https://github.com/zephyrproject-rtos/zephyr/pull/82184

rruuaanng commented on tests/kernel/mem_protect/stackprot/src/main.c:
```suggestion
	if (should_fail == 0) {
```
(Forgive me for being mobile and can't see more context.)

ceolin replied:
I am curious about this suggestion. `should_fail` is a boolean type, so AFAIK `!should_fail` is a valid construction according with our and MISRA-C code guidelines. In other hand, comparing boolean with integer is probably violating MISRA-C.

rruuaanng replied:
Oh, you can ignore my suggestion. I mistakenly thought it was int :)

ruuddw commented:
As described in the issue, there may be 4 levels of protection (```
-fstack-protector, -fstack-protector-strong, -fstack-protector-all and
-fstack-protector-explicit
```). The PR excludes the lowest level (`-fstack-protector`), any reason? Isn't it confusing to take over the post-fix names but change the no-postfix version to '-all'?

ruuddw commented:
Thanks for including the ARC Metaware compiler flags. FYI, that follows the LLVM/Clang approach - no `-explicit`. Note that the reverse `(__attribute__((no_stack_protector))` is supported, which ensures no canary is inserted even if the heuristics would have decided to add one.

cfriedt reviewed (commented):
Looks good to me - I think I just spotted a small typo

tejlmand commented on kernel/Kconfig:
```diff
@@ -862,25 +862,47 @@ config XIP
-	bool "Compiler stack canaries"
-	depends on ENTROPY_GENERATOR || TEST_RANDOM_GENERATOR
-	select NEED_LIBC_MEM_PARTITION if !STACK_CANARIES_TLS
+config NEED_STACK_CANARIES
```
can we use `REQUIRES_STACK_CANARIES` to stay aligned with other Zephyr toolchain / build system settings, like
`REQUIRES_STD_C99`, `REQUIRES_FULL_LIBC`, etc.

ceolin replied:
Done :)

tejlmand commented on kernel/Kconfig:
```diff
@@ -862,25 +862,47 @@ config XIP
+	bool
 	help
-	  This option enables compiler stack canaries.
+	  Hidden option to signal that software stack protection is enabled.
```
```suggestion
	  Hidden option to signal that software stack protection is required.
```

ceolin replied:
done

tejlmand commented on kernel/Kconfig:
perhaps shorten the names as those are choice entries:
Current:
![image](https://github.com/user-attachments/assets/c7523824-95f8-42ba-9eb6-2cc08cb893c4)

Proposal:
![image](https://github.com/user-attachments/assets/b8f390bc-e087-4233-923a-b432a0158d0c)

```suggestion
	bool "All functions"
```
Only a proposal, thus non-blocking .

ceolin replied:
Applied suggestions, much better indeed.

tejlmand commented on kernel/Kconfig:
related to other Kconfig choice entry prompt proposal:

```suggestion
	bool "Most functions"
```
Only a proposal, thus non-blocking .

tejlmand commented on kernel/Kconfig:
```diff
@@ -900,6 +900,15 @@ config STACK_CANARIES_STRONG
 	  frame addresses.
 
+config STACK_CANARIES_EXPLICIT
+	bool "Compiler stack canaries in explicit functions"
```
related to other Kconfig choice entry prompt proposal:

```suggestion
	bool "Explicit functions"
```
Only a proposal, thus non-blocking .

tejlmand reviewed (changes requested):
Looks good.

Some minor suggestions, the main one being the proposal to use `REQUIRES` instead of `NEED`.

ceolin commented on tests/kernel/mem_protect/stackprot/src/main.c:
fixed, thanks for spotting it.

vujjuri commented:
> As described in the issue, there may be 4 levels of protection (`-fstack-protector, -fstack-protector-strong, -fstack-protector-all and -fstack-protector-explicit`). The PR excludes the lowest level (`-fstack-protector`), any reason? Isn't it confusing to take over the post-fix names but change the no-postfix version to '-all'?

I agree here. All the 4 levels are different. User should have provision to choose among 4 options

vujjuri commented on cmake/compiler/arcmwdt/compiler_flags.cmake:
```diff
@@ -168,6 +168,8 @@ set_compiler_property(PROPERTY imacros -imacros)
 set_compiler_property(PROPERTY security_canaries -fstack-protector-all)
+set_compiler_property(PROPERTY security_canaries_strong -fstack-protector-strong)
+set_compiler_property(PROPERTY security_canaries_explicit -fstack-protector-explicit)
 
```
`-fstack-protector-all` and `-fstack-protector` are different options. User should be able to select them different. If no explicit choice is specified by user, `-fstack-protector-all` can be taken to make it compatibility with previous versions

vujjuri commented:
> As described in the issue, there may be 4 levels of protection (`-fstack-protector, -fstack-protector-strong, -fstack-protector-all and -fstack-protector-explicit`). The PR excludes the lowest level (`-fstack-protector`), any reason? Isn't it confusing to take over the post-fix names but change the no-postfix version to '-all'?

I agree. `-fstack-protector-all` and `-fstack-protector` are different options. User should be able to select among those. If no explicit choice is specified by user, `-fstack-protector-all` can be taken to make it compatibility with previous versions

ceolin commented:
> > As described in the issue, there may be 4 levels of protection (`-fstack-protector, -fstack-protector-strong, -fstack-protector-all and -fstack-protector-explicit`). The PR excludes the lowest level (`-fstack-protector`), any reason? Isn't it confusing to take over the post-fix names but change the no-postfix version to '-all'?
> 
> I agree. `-fstack-protector-all` and `-fstack-protector` are different options. User should be able to select among those. If no explicit choice is specified by user, `-fstack-protector-all` can be taken to make it compatibility with previous versions


Yep, they are not the same, I kept it to keep the same the symbol name and behavior, otherwise I would need to change the current symbol to `-fstack-protector` to add `STACK_CANARIES_ALL` for he current option.

That's said I have no hard feeling about adding it, I just don't know how much we gain adding it.

ruuddw commented:
> That's said I have no hard feeling about adding it, I just don't know how much we gain doing it.

My preference would be to align the Zephyr names with the compiler names. I find it confusing to map the less intrusive "STACK_CANARIES" to the strongest option "stack-protector-all". And to have "STACK_CANARIES_STRONG" to be weaker than "STACK_CANARIES".

For the explicit option, unfortunately there is a mismatch between gcc and llvm. From what I understand gcc uses pragma's to explicitly enable canaries for a function, and llvm/clang has pragma's to explicitly disable canaries. I don't see an easy way to abstract from this and have a unified approach.

ruuddw commented on cmake/compiler/arcmwdt/compiler_flags.cmake:
```diff
@@ -168,6 +168,8 @@ set_compiler_property(PROPERTY imacros -imacros)
 #no support of -mstack-protector-guard=global"
 set_compiler_property(PROPERTY security_canaries -fstack-protector-all)
+set_compiler_property(PROPERTY security_canaries_strong -fstack-protector-strong)
+set_compiler_property(PROPERTY security_canaries_explicit -fstack-protector-explicit)
```
MetaWare is like llvm/clang, there is no "-fstack-protector-explicit" option, and instead of `__attribute__((stack_protector))`  the ` __attribute__((no_stack_protector))` is used for explicit control.

ceolin replied:
So it is the other way around, instead of tells which functions the compiler should instrument you tells which ones shouldn't ?

ruuddw commented on include/zephyr/toolchain/llvm.h:
```diff
@@ -28,6 +28,13 @@
+ * Define the macro before including gcc.h to avoid
+ * problems with non existent attribute.
+ */
+#define __stack_protect
```
Something similar will be needed for MetaWare as well.

ceolin committed:
security: Add default stack protection level

STACK_CANARIES was enabling canaries in all functions using the compiler
flag -fstack-protector-all. This became confuse with the addition of the
options STRONG and EXPLICIT.

This commit adds the missing option (default level) and disambiguous the
options mapping them close to the compiler flags.

Now we have the following options:

STACK_CANARIES            -> fstack-protector
STACK_CANARIES_STRONG     -> fstack-protector-strong
STACK_CANARIES_ALL        -> fstack-protector-all
STACK_CANARIES_EXPLICIT   -> fstack-protector-explicit

Note that from now on STACK_CANARIES_ALL is the symbol that adds canaries
for all functions.

Signed-off-by: Flavio Ceolin <flavio.ceolin@gmail.com>

ceolin committed:
doc: migration-guide-4.1: Information about stack canaries

Add information about the changing in the stack canaries symbol behavior.

Signed-off-by: Flavio Ceolin <flavio.ceolin@gmail.com>

ceolin commented:
> > That's said I have no hard feeling about adding it, I just don't know how much we gain doing it.
> 
> My preference would be to align the Zephyr names with the compiler names. I find it confusing to map the less intrusive "STACK_CANARIES" to the strongest option "stack-protector-all". And to have "STACK_CANARIES_STRONG" to be weaker than "STACK_CANARIES".
> 
> For the explicit option, unfortunately there is a mismatch between gcc and llvm. From what I understand gcc uses pragma's to explicitly enable canaries for a function, and llvm/clang has pragma's to explicitly disable canaries. I don't see an easy way to abstract from this and have a unified approach.

I can add macros for both, disable / enable. There is no internal usage of these macros and we can enforce that only application uses it. The application will need to know which toolchain is been used but there is not much else we can do. These compilers have different approach for this.

ceolin commented:
> `I can add macros for both, disable / enable. There is no internal usage of these macros and we can enforce that only application uses it. The application will need to know which toolchain is been used but there is not much else we can do. These compilers have different approach for this.`
> 
> That could work, but it doesn't really abstract to a unified solution. Maybe just using the attributes instead of a macro is good enough. Risk of macros that get compiled away is that people may assume a canary is inserted or prevented while it isn't.

That is a good point indeed. I will remove it.

ruuddw commented:
Thanks, just a final question to check if it was a design choice or mistake. MetaWare doesn't support `-fstack-protector-explicit`, selecting it will result in compiler warnings. An alternative to the compiler warning could be a configuration time solution blocking the choice, or a pre-processor #error instead of compiler warning. In any case the user will notice, so I don't have a strong preference. Only concern now is that the user doesn't know if this is a bug or a feature. So I suggest to at least add a remark in the description of `config STACK_CANARIES_EXPLICIT` that it is not supported for all toolchains.

ceolin commented:
> Thanks, just a final question to check if it was a design choice or mistake. MetaWare doesn't support `-fstack-protector-explicit`, selecting it will result in compiler warnings. An alternative to the compiler warning could be a configuration time solution blocking the choice, or a pre-processor #error instead of compiler warning. In any case the user will notice, so I don't have a strong preference. Only concern now is that the user doesn't know if this is a bug or a feature. So I suggest to at least add a remark in the description of `config STACK_CANARIES_EXPLICIT` that it is not supported for all toolchains.

I will fix it, @ruuddw where can I find the MetaWare compiler documentation ? I briefly tried to find this yesterday but failed.

Thanks for checking it closely !

ruuddw commented:
There's no online documentation for MetaWare, as far as I know. There is a free version, called [MetaWare Lite](https://www.synopsys.com/cgi-bin/arcmwtk_lite/reg1.cgi). Windows only, and image size restricted to 32 kB (good to ensure people care about codesize...). That includes documentation. Appreciate your efforts - it is also fine to rely on the ARC team and me to provide the inputs and testing. We have internal CI running with MetaWare as well to catch problems.

ruuddw reviewed (approved):
All good now!
